File: Assignment_1/task3.scala
import java.io.{File, PrintWriter}

import org.apache.spark.HashPartitioner
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.json4s._
import org.json4s.jackson.JsonMethods._

object task3 {
  case class Review_Class(review_id: String, user_id: String, business_id: String, stars: String, text: String, date: String)

  def main(args: Array[String]) {

    val sparkSession = SparkSession.builder().appName("task3").config("spark.master", "local[*]").getOrCreate()

    val spark_context = sparkSession.sparkContext
    spark_context.setLogLevel("ERROR")
    val inputFile = args(0)
    val outputFile = args(1)
    val partition_type = args(2)
    val n_partitions = args(3)
    val n = args(4)
    val rdd = spark_context.textFile(inputFile)
      .map(extractData)
      .map(x => (x.business_id, x.date, x.review_id, x.stars, x.text, x.user_id))
    var rdd_reqd = rdd.map(x => (x._1, 1))
    // custimized partition
    if (partition_type == "customized") {
      rdd_reqd = rdd_reqd.partitionBy(new HashPartitioner(Integer.parseInt(n_partitions)))
    }

    val n_partition = rdd_reqd.getNumPartitions
    val items = calc_items(rdd_reqd)
    val result = calc_more_than_n(rdd_reqd, Integer.parseInt(n))

    println(n_partition, items.mkString(","), result.mkString(","))
    process_output(n_partition, items, result, outputFile)
  }

  def calc_items(rdd: RDD[(String, Int)]): Array[Int] = {
    val rdd_rq = rdd.glom().map(_.length)
    return rdd_rq.collect()
  }

  def process_output(partA: Long, partB: Array[Int], partC: Array[(String, Int)], output: String) = {
    val writerObject = new PrintWriter(new File(output))
    writerObject.write('{')
    writerObject.write("\"n_partitions\": " + partA + ", ")
    writerObject.write("\"n_items\": [")
    for (i <- partB) {
      writerObject.write(i.toString)
      if (partB.indexOf(i) < partB.length - 1) {
        writerObject.write(", ")
      }
    }
    writerObject.write("], ")
    writerObject.write("\"result\": [")
    for (i <- partC) {
      writerObject.write("[\"" + i._1 + "\", " + i._2 + "]")
      if (partC.indexOf(i) < partC.length - 1) {
        writerObject.write(",")
      }
    }
    writerObject.write("]}")
    writerObject.close()
  }

  def calc_more_than_n(rdd: RDD[(String, Int)], n: Int): Array[(String, Int)] = {
    val rdd_red = rdd.reduceByKey(_ + _)
    val rdd_cal = rdd_red.filter(x => x._2 > n)
    return rdd_cal.collect()
  }

  def extractData(row: Any): Review_Class = {
    implicit val formats: DefaultFormats.type = DefaultFormats
    parse(row+"").extract[Review_Class]
  }
}

File: Assignment_1/task3.py
from pyspark import SparkContext
import sys
import time
import json

start_time = time.time()

input_file = ''
output_file = ''
partition_type = ''
n_partitions = 0
n = 0
ans = {}
class Building_Structure():

	def read_input(self):
		global input_file, output_file, partition_type, n_partitions, n
		input_file = sys.argv[1]
		output_file = sys.argv[2]
		partition_type = sys.argv[3]
		n_partitions = sys.argv[4]
		n = sys.argv[5]

	def process_output(self):
		# print(ans)
		f = open(output_file,"w")
		json.dump(ans,f)
		f.close()

	def convert_to_list(self, s):
		review_list = []
		for i in s:
			review_list.append(list(i))
		return review_list

def partition_function(key):
	#taking hash value of the business_id and taking mod with n_partitions
	return hash(key)%int(n_partitions)

def process(row, n):
	return row[1] > n

def calc_items(rdd_columns):
	rdd_columns_modified = rdd_columns.glom().map(len)
	ans['n_items'] = rdd_columns_modified.collect()

def calc_more_than_n(rdd_columns):
	rdd_columns_modified = rdd_columns.reduceByKey(lambda a,b :a+b)
	rdd_columns_crop = rdd_columns_modified.filter(lambda row:process(row, int(n)))
	ans['result'] = Building_Structure().convert_to_list(rdd_columns_crop.collect())

def main():

	building_structure = Building_Structure()

	building_structure.read_input()
	# print("Arguments Passed", input_file, output_file, partition_type, n_partitions, n)

	#creating Object
	sc_object = SparkContext()
	rdd_lines = sc_object.textFile(input_file).map(lambda line: json.loads(line))
	rdd_columns_reqd = rdd_lines.map(lambda row: (row['business_id'], row['review_id']))
	#key value pairs
	rdd_columns = rdd_columns_reqd.map(lambda data_row: (data_row[0],1))
	if partition_type == "customized":
		rdd_columns = rdd_columns.partitionBy(int(n_partitions), partition_function)

	ans['n_partitions'] = rdd_columns.getNumPartitions()
	calc_items(rdd_columns)
	calc_more_than_n(rdd_columns)
	building_structure.process_output()


if __name__ == "__main__":
	main()

# print("--- %s seconds ---" % (time.time() - start_time))